### PR TITLE
#225 Add secret link to download ics calendar

### PR DESCRIPTION
**Pull request by asmaNode**
BACKEND PR : https://ci.linagora.com/linagora/lgs/openpaas/linagora.esn.calendar/merge_requests/698/
**DEMO with connected user** : https://streamable.com/5cf1bk
**DEMO with disconnected user** : https://streamable.com/megfwj
**Full demo with reset secret link** : https://streamable.com/vebw0b
What have bee done :

- [x] Generate a token for secret link

- [x] Create new endpoint to download calendar in  ics format 
- [x] Reset token to disable the old links

### CODE REVIEW

File: src/esn.calendar.libs/app/services/calendar-service.js [hunk 1]
```diff
@@ -371,5 +375,21 @@ require('./calendar-cache.js');
         FileSaver.saveAs(calendarBlob, `${calendarId}.ics`);
       });
     }
+    /**
+      * Get the generated token for the secret link
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-service.js:379]:
```suggestion
    }

    /**
      * Get the generated token for the secret link
```

File: src/esn.calendar.libs/app/services/calendar-service.js [hunk 2]
```diff
@@ -371,5 +375,21 @@ require('./calendar-cache.js');
...
+      * @param {Object} jwtPayload The jwtPayload that contains the information to generate token
+      * @return {String} a generated token with calendarHomeId and calendar id payload
+      */
+    function generateTokenForSecretLink(jwtPayload) {
+      return $http({ method: 'POST', url: getGenerateJWTforSecretLinkUrl(), data: jwtPayload });
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-service.js:384]: 1. We don't handle sending requests in `services/calendar-service.js`. It should behandled it `services/calendar-api.js`.
2. We should be using `calendarRestangular` wherever possible instead of `$http` because we have some configurations for it.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 3]
```diff
@@ -110,5 +111,21 @@ require('../../components/modals/calendar-delete-confirmation/calendar-delete-co
...
     }
+    function _createSecretLinkkwithToken() {
+      const jwtPayload = {
+        calendarHomeId: self.calendarHomeId,
+        calendarId: self.calendar.id,
+        userId: session.user._id
+
+      };
+
+      calendarService.generateTokenForSecretLink(jwtPayload)
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:122]: This should return the token directly, not the response body.

File: src/esn.calendar.libs/app/services/calendar-service.js [hunk 4]
```diff
@@ -371,5 +373,15 @@ require('./calendar-cache.js');
...
+
+    /**
+      * Get the generated token for the secret link
+      * @param {Object} jwtPayload The jwtPayload that contains the information to generate token
+      * @return {String} a generated token with calendarHomeId and calendar id payload
+      */
+    function generateTokenForSecretLink(jwtPayload) {
+      return calendarAPI.generateToken(jwtPayload).then(token => token);
+
+    }
```
Comment (rezk2ll) [inline src/esn.calendar.libs/app/services/calendar-service.js:385]:
```suggestion
    function generateTokenForSecretLink(jwtPayload) {
      return calendarAPI.generateToken(jwtPayload);
    }
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug [hunk 5]
```diff
@@ -52,6 +52,18 @@
...
       esn-clipboard-url(url="$ctrl.caldavurl")
+  .calendar-link(ng-if="$ctrl.calendarSecretLink")
+    hr
+    .settings-label
+      i.mdi.mdi-link
+      span {{ 'The secret link' | translate }}
+    .settings-value
+      esn-clipboard-url(url="$ctrl.calendarSecretLink")
+    .settings-text
+      span {{'Use this address to access the iCalendar format file (.ics). Warning: It allows anyone with the address to view your calendar and all of its events.' | translate}}
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:63]:
```suggestion
      span {{'Anyone can export this calendar using this link.' | translate}}
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug [hunk 6]
```diff
@@ -52,6 +52,18 @@
       span {{ 'CalDAV URL' | translate }}
     .settings-value
       esn-clipboard-url(url="$ctrl.caldavurl")
+  .calendar-link(ng-if="$ctrl.calendarSecretLink")
+    hr
+    .settings-label
+      i.mdi.mdi-link
+      span {{ 'The secret link' | translate }}
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:59]:
```suggestion
      span {{ 'Shareable calendar link' | translate }}
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug [hunk 7]
```diff
@@ -52,6 +52,18 @@
...
+    .settings-value
+      esn-clipboard-url(url="$ctrl.calendarSecretLink")
+    .settings-text
+      span {{'Use this address to access the iCalendar format file (.ics). Warning: It allows anyone with the address to view your calendar and all of its events.' | translate}}
+    .settings-actions
+      a.btn.btn-primary.btn-sm(ng-href="{{::$ctrl.calendarSecretLink}}") {{ 'Reset secret address' | translate }}
+    .settings-actions(ng-if="$ctrl.canDeleteCalendar")
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:66]:
```suggestion
```

File: src/esn.calendar.libs/i18n/zh.json [hunk 8]
```diff
@@ -278,14 +290,5 @@
...
-  "The event has not been found": "该事件尚未找到",
-  "An error occured on our side, please contact administrator if problem persists": "我们发生错误，如果问题仍然存在，请与管理员联系",
-  "Search in events": "搜索活动",
-  "Forwarding this invitation could allow any recipient to send a response to the organizer and be added to the guest list, or invite others regardless of their own invitation status, or to modify your RSVP.": "转发此邀请后，任何收件人都可以向组织者发送回复，并且被添加到邀请对象列表；无论其邀请状态为何，他们都可以邀请其他人；或者，他们也可以修改您的回复。",
-  "Saving configuration...": "保存设置…",
-  "Configuration saved": "设置保存成功",
-  "Failed to save configuration": "设置保存失败"
-}
+  "Your proposal has been sent": "您的建议已发送"
+}
```
Comment (rezk2ll) [inline src/esn.calendar.libs/i18n/zh.json:294]:
```suggestion
}

```

File: src/esn.calendar.libs/i18n/vi.json [hunk 9]
```diff
@@ -278,14 +290,5 @@
...
-  "The event has not been found": "Sự kiện chưa được tìm thấy",
-  "An error occured on our side, please contact administrator if problem persists": "Đã xảy ra lỗi ở phía chúng tôi, vui lòng liên hệ với quản trị viên nếu sự cố vẫn tiếp diễn",
-  "Search in events": "Tìm sự kiện",
-  "Forwarding this invitation could allow any recipient to send a response to the organizer and be added to the guest list, or invite others regardless of their own invitation status, or to modify your RSVP.": "Nhờ chuyển tiếp lời mời này, mọi người nhận có thể gửi câu trả lời cho người tổ chức và được thêm vào danh sách khách hoặc mời những người khác bất kể trạng thái lời mời của họ hoặc sửa đổi câu trả lời.",
-  "Saving configuration...": "Đang lưu cấu hình...",
-  "Configuration saved": "Đã lưu cấu hình",
-  "Failed to save configuration": "Không thể lưu cấu hình"
-}
+  "Your proposal has been sent": "Đề xuất của bạn đã được gửi đi"
+}
```
Comment (rezk2ll) [inline src/esn.calendar.libs/i18n/vi.json:294]:
```suggestion
}

```

File: src/esn.calendar.libs/app/services/calendar-service.js [hunk 10]
```diff
@@ -25,6 +25,7 @@ require('./calendar-cache.js');
     userUtils,
     FileSaver,
     CAL_EXPORT_FILE_TYPE
+
```
Comment (rezk2ll) [inline src/esn.calendar.libs/app/services/calendar-service.js:28]:
```suggestion
```

File: src/esn.calendar.libs/app/services/calendar-api.js [hunk 11]
```diff
@@ -316,5 +317,21 @@ require('./http-response-handler.js');
...
         .then(calHttpResponseHandler(200, _.property('data')));
     }
+    /**
+      * POST resuest used to generate token for secret link to download ics calendar
+      * @param {Object} jwtPayload   paylod to generate token
+      */
+    function generateToken(jwtPayload) {
+
+      return calendarRestangular.one('calendars/generateJWTforSecretLink').customPOST(jwtPayload)
+        .then(calHttpResponseHandler(200, _.property(['data'])))
```
Comment (rezk2ll) [inline src/esn.calendar.libs/app/services/calendar-api.js:327]:
```suggestion
        .then(calHttpResponseHandler(200, _.property('data')))
```
does this work without the `[ ]`?

**Review by rezk2ll — APPROVED** [review]

File: src/esn.calendar.libs/app/services/calendar-api.spec.js [hunk 12]
```diff
@@ -643,6 +643,22 @@ describe('The calendar module apis', function() {
...
+      it('should get the calendar export path', function() {
+        const expectedResponse = 'some vcal stuff';
+        const jwtPaylod = {
+          calendarHomeId: 'calendarHomeId',
+          calendarId: 'calendarId',
+          userId: 'userId'
+        };
+
+        this.$httpBackend.expectPOST('/calendar/api/calendars/generateJWTforSecretLink').respond(expectedResponse);
+        this.calendarAPI.generateToken(jwtPaylod);
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-api.spec.js:657]:
```suggestion
        this.calendarAPI.generateToken(jwtPayload);
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 13]
```diff
@@ -110,5 +111,19 @@ require('../../components/modals/calendar-delete-confirmation/calendar-delete-co
...
+        .then(resp => {
+          self.calendarSecretLink = `${window.location.origin}/calendar/api/calendars/secretLink?jwt=${resp.token}`;
+        });
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:125]: The service should return something like this:

```suggestion
        .then(token => {
          self.calendarSecretLink = `${window.location.origin}/calendar/api/calendars/secretLink?jwt=${token}`;
        });
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug (hunk 5 above)
Comment (huy-ta) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:63]: Let's keep the text like that. "export" is another feature.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js [hunk 14]
```diff
@@ -469,6 +478,25 @@ describe('The calendar configuration tab delegation controller', function() {
       expect(CalendarConfigurationTabMainController.sharedCalendarOwner).to.equal(getOwnerResult);
       expect(CalendarConfigurationTabMainController.displayNameOfSharedCalendarOwner).to.equal(userUtilsResult);
     });
+    it('should create secret link with token to donwload ics calendar', function(done) {
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js:481]:
```suggestion
    it('should create secret link with token to download ics calendar', function(done) {
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 15]
```diff
@@ -110,5 +111,19 @@ require('../../components/modals/calendar-delete-confirmation/calendar-delete-co
     function exportCalendar() {
       calendarService.exportCalendar(self.calendarToExport.calendarHomeId, self.calendarToExport.id);
     }
+    function _createSecretLinkkwithToken() {
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:114]:
```suggestion
    }

    function _createSecretLinkkwithToken() {
```

Comment (rezk2ll) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:114]:
```suggestion
    }
    function _createSecretLinkwithToken() {
```

File: src/esn.calendar.libs/app/services/calendar-api.js (hunk 11 above)
Comment (asmaNode) [reply src/esn.calendar.libs/app/services/calendar-api.js:327]: yes , it is working , in the first i want to fetch a nested object forthat i used an array  like this  _.property(['data', 'token'] but got undefined

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug (hunk 5 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:63]: i used the text mentionned in description

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug (hunk 6 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:59]: the text is in the spec

File: src/esn.calendar.libs/app/services/calendar-service.js [hunk 16]
```diff
@@ -371,5 +372,14 @@ require('./calendar-cache.js');
...
       });
     }
+
+    /**
+      * Get the generated token for the secret link
+      * @param {Object} jwtPayload The jwtPayload that contains the information to generate token
+      * @return {String} a generated token with calendarHomeId and calendar id payload
+      */
+    function generateTokenForSecretLink(jwtPayload) {
+      return calendarAPI.generateToken(jwtPayload).then(resp => resp.token);
```
Comment (rezk2ll) [inline src/esn.calendar.libs/app/services/calendar-service.js:382]:
```suggestion
      return calendarAPI.generateToken(jwtPayload).then(({ token }) => token);
```

**Review by rezk2ll — DISMISSED** [review]

File: src/esn.calendar.libs/app/services/calendar-api.js [hunk 17]
```diff
@@ -316,5 +317,21 @@ require('./http-response-handler.js');
       return calDavRequest('get', calendarPath, { Accept: CAL_ACCEPT_EXPORT_HEADER })
         .then(calHttpResponseHandler(200, _.property('data')));
     }
+
+    /**
+     * POST resuest used to generate token for secret link to download ics calendar
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-api.js:322]:
```suggestion
     * POST request used to generate token for the secret link to download a calendar in iCal format (*.ics)
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 18]
```diff
@@ -18,15 +18,18 @@ require('../../components/modals/calendar-delete-confirmation/calendar-delete-co
...
     calUIAuthorizationService,
-    calCalendarDeleteConfirmationModalService
+    calCalendarDeleteConfirmationModalService,
+    calCalendarSecretAddressConfirmationModalService
   ) {
     var self = this;
 
     self.$onInit = $onInit;
     self.openDeleteConfirmationDialog = openDeleteConfirmationDialog;
+    self.openGetSecretLinkFConfirmationDialog = openGetSecretLinkFConfirmationDialog;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:28]:
```suggestion
    self.openGetSecretLinkConfirmationDialog = openGetSecretLinkConfirmationDialog;
```

File: src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
+(function(angular) {
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js:1]: No need for IIFE anymore.

File: src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.pug [hunk 20]
```diff
@@ -0,0 +1,10 @@
+.modal.secret-address-confirmation
+  .modal-dialog.modal-sm
+    .modal-content
+      .modal-header
+        h4.modal-title {{ 'Get a new secret address?' | translate }}
+      .modal-body
+        span {{ 'This will make the current secret address invalid(if any).Are you sure want to generate a new secret address? ' | translate:ctrl }}
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.pug:7]:
```suggestion
        span {{ 'This will make the current secret address invalid (if any). Are you sure want to generate a new secret address? ' | translate }}
```

Comment (huy-ta) [reply src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.pug:7]: It is also missing translation.

File: src/esn.calendar.libs/app/services/calendar-api.js [hunk 21]
```diff
@@ -316,5 +317,21 @@ require('./http-response-handler.js');
...
+     * @param {Object} jwtPayload   paylod to generate token
+     */
+    function generateToken(jwtPayload) {
+
+      return calendarRestangular.one('calendars/generateJWTforSecretLink').customPOST(jwtPayload)
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-api.js:327]:
```suggestion
      return calendarRestangular.one('calendars').one('generateJWTforSecretLink').customPOST(jwtPayload)
```

File: src/esn.calendar.libs/app/services/calendar-api.js [hunk 22]
```diff
@@ -316,5 +317,21 @@ require('./http-response-handler.js');
...
+        .then(calHttpResponseHandler(200, _.property('data')))
+        .catch(function(error) {
+          notificationFactory.weakError('Failed to generate token for secret', 'Cannot join the server, please try later');
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-api.js:330]:
```suggestion
          notificationFactory.weakError('Failed to generate token for the secret address', 'Cannot join the server, please try later');
```

Just `for secret` is too ambiguous for users.

File: src/esn.calendar.libs/app/services/calendar-api.spec.js [hunk 23]
```diff
@@ -643,6 +643,22 @@ describe('The calendar module apis', function() {
         this.$httpBackend.flush();
       });
     });
+
+    describe('the generate token for secret link request', function() {
+      it('should get the calendar export path', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-api.spec.js:648]: You are not testing really testing the function here.

You need to write tests that make sure any changes to the code's logic will break the tests.

1. There is no testing for this part:

```js
.then(calHttpResponseHandler(200, _.property('data')))
```

2. There is no testing for this part:

```js
.catch(function(error) {
   notificationFactory.weakError('Failed to generate token for secret', 'Cannot join the server, please try later');

   return $q.reject(error);
});
```

File: src/esn.calendar.libs/app/services/calendar-service.spec.js [hunk 24]
```diff
@@ -1141,4 +1141,19 @@ describe('The calendarService service', function() {
       expect(fileSaveMock.saveAs).to.have.calledWith(expectedBlob, 'calendarId.ics');
     });
   });
+
+  describe('the generateTokenForSecretLink function', function() {
+    it('should call the calendarAPI generate jwt', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-service.spec.js:1146]:
```suggestion
    it('should call calendarAPI.getToken to generate a token with correct params', function() {
```

File: src/esn.calendar.libs/app/services/calendar-service.spec.js [hunk 25]
```diff
@@ -1141,4 +1141,19 @@ describe('The calendarService service', function() {
...
+      const jwtPayload = {
+        calendarHomeId: 'calendarHomeId',
+        calendarId: 'calendarId',
+        userId: 'userId'
+      };
+
+      this.calendarAPI.generateToken = sinon.stub().returns($q.when());
+      this.calendarService.generateTokenForSecretLink(jwtPayload);
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/calendar-service.spec.js:1154]: You forgot to test this part:

```js
      .then(({ token }) => token);
```

File: src/esn.calendar.libs/app/services/calendar-api.js (hunk 22 above)
Comment (huy-ta) [reply src/esn.calendar.libs/app/services/calendar-api.js:330]: Also no translation.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 26]
```diff
@@ -18,15 +18,18 @@ require('../../components/modals/calendar-delete-confirmation/calendar-delete-co
...
     self.removeCalendar = removeCalendar;
     self.unsubscribe = unsubscribe;
     self.exportCalendar = exportCalendar;
+    self.createSecretLinkwithToken = createSecretLinkwithToken;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:32]: There is no need for this.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js [hunk 27]
```diff
@@ -469,6 +489,25 @@ describe('The calendar configuration tab delegation controller', function() {
       expect(CalendarConfigurationTabMainController.sharedCalendarOwner).to.equal(getOwnerResult);
       expect(CalendarConfigurationTabMainController.displayNameOfSharedCalendarOwner).to.equal(userUtilsResult);
     });
+
+    it('should create secret link with token to donwload ics calendar', function(done) {
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js:493]: Shouldn't this be in a separate `describe` block?

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js [hunk 28]
```diff
@@ -469,6 +489,25 @@ describe('The calendar configuration tab delegation controller', function() {
...
+      const jwtPayload = {
+        calendarHomeId: 'calendarHomeId',
+        calendarId: 'calendarId',
+        userId: 'userid'
+      };
+
+      CalendarConfigurationTabMainController.$onInit();
+      $rootScope.$digest();
+      calendarService.generateTokenForSecretLink(jwtPayload)
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js:502]: You are not supposed to be testing `calendarService.generateTokenForSecretLink` here, so you shouldn't call it here.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js [hunk 29]
```diff
@@ -469,6 +489,25 @@ describe('The calendar configuration tab delegation controller', function() {
...
+        .then(() => {
+          expect(calendarService.generateTokenForSecretLink).to.have.been.called;
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.spec.js:504]: This is equivalent to `expect(true).to.be.true`.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug [hunk 30]
```diff
@@ -52,6 +52,17 @@
       span {{ 'CalDAV URL' | translate }}
     .settings-value
       esn-clipboard-url(url="$ctrl.caldavurl")
+  .calendar-link
+    hr
+    .settings-label
+      i.mdi.mdi-link
+      span {{ 'Secret address in iCal format' | translate }}
+    .settings-value
+      esn-clipboard-url(ng-if="$ctrl.calendarSecretLink")(url="$ctrl.calendarSecretLink")
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:61]:
```suggestion
      esn-clipboard-url(ng-if="$ctrl.calendarSecretLink", url="$ctrl.calendarSecretLink")
```

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug [hunk 31]
```diff
@@ -52,6 +52,17 @@
...
+    .settings-text
+      span {{'Use the following button to get the address to access the iCalendar format file (.ics). Warning: It allows anyone with the address to view your calendar and all of its events.' | translate}}
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.pug:63]:
```suggestion
      span {{ 'Use the following button to get the address to access the iCalendar format file (.ics). Warning: It allows anyone with the address to view your calendar and all of its events.' | translate }}
```

File: src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js [hunk 32]
```diff
@@ -0,0 +1,21 @@
...
+  'use strict';
+
+  angular.module('esn.calendar')
+    .factory('calCalendarSecretAddressConfirmationModalService', calCalendarSecretAddressConfirmationModalService);
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js:5]: Please make sure you test this in production mode.

**Comment by huy-ta** [comment]
Well since we don't have the reset feature yet, you should skip displaying the dialog. It's better that way than lying to our users that the previous secret links will be invalidated (but in fact they won't).

**Comment by huy-ta** [comment]
And here's a tip for writing unit tests: When you change something in your code and your unit tests don't fail, it probably means that your unit tests are not correct/sufficient.

**Comment by huy-ta** [comment]
And I recommend you turn off your auto format on save setting. It's changing a lot of things that are not related to your PRs and it makes it harder to review. Also it's adding some weird spaces/blank lines.

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js (hunk 26 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:32]: must be defined for test

File: src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js (hunk 32 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/components/modals/calendar-secret-address-confirmation/calendar-secret-address-confirmation-modal.service.js:5]: yes , it is working

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js (hunk 26 above)
Comment (huy-ta) [reply src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js]: You don't need it defined for the tests. And you shouldn't do that most of the time. The code should not be changed only for the purpose of testing because it will cause confusion later on. For example, when you do this: `self.createSecretLinkwithToken = createSecretLinkwithToken;`, people expect that it will be used in the templates.

By the way, this is not the function you need to test. The function you need to test here is `openGetSecretLinkConfirmationDialog`. Right now with your tests someone can remove the `openGetSecretLinkConfirmationDialog` function and break the feature, but the tests will still pass.

So here's how it should be done:

With this part of code:

```js
function openGetSecretLinkFConfirmationDialog() {
  calCalendarSecretAddressConfirmationModalService(self.calendar, createSecretLinkwithToken);
}
```

You can do something like: 

```
calCalendarDeleteConfirmationModalServiceMock = sinon.stub();

// ...

$provide.value('calCalendarDeleteConfirmationModalServiceMock', calCalendarDeleteConfirmationModalServiceMock)

// And in the tests:

expect(calCalendarDeleteConfirmationModalServiceMock).to.have.been.calledOnce;
expect(calCalendarDeleteConfirmationModalServiceMock.getCall(0).args[0]).to.equal(CalendarConfigurationTabMainController.calendar);

const createSecretLinkWithTokenFn = calCalendarDeleteConfirmationModalServiceMock.getCall(0).args[1];

createSecretLinkWithTokenFn();

expect(calendarServiceMock.generateTokenForSecretLink).to.have.been.calledWith(...);

//
```

Perhaps we can do a pair programming session after the daily today.

**Review by rezk2ll — APPROVED** [review]

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 33]
```diff
@@ -1,119 +1,138 @@
...
+  }
 
-    function canExportIcs() {
-      return !self.newCalendar && calUIAuthorizationService.canExportCalendarIcs(self.calendar, session.user._id);
-    }
+  function removeCalendar() {
+    calendarService.removeCalendar(self.calendarHomeId, self.calendar).then(function() {
+      $state.go('calendar.main');
+    });
+  }
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:90]: I smell something fishy here. Why is this function added? We don't do anything about the calendar removal functionality right?

File: src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js [hunk 34]
```diff
@@ -1,119 +1,138 @@
...
 
-    function _canModifyPublicSelection() {
-      return self.newCalendar || calUIAuthorizationService.canModifyPublicSelection(self.calendar, session.user._id);
-    }
+  function canDeleteCalendar() {
+    return !self.newCalendar && calUIAuthorizationService.canDeleteCalendar(self.calendar, session.user._id);
+  }
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar-configuration/calendar-configuration-tab-main/calendar-configuration-tab-main.controller.js:93]: This is weird too, why did you add this function?

**Comment by huy-ta** [comment]
Your last commit seems to be really weird. Can you please put the IIFE removal into a separate commit?

**Review by rezk2ll — DISMISSED** [review]

**Review by rezk2ll — DISMISSED** [review]

**Review by huy-ta — APPROVED** [review]

**Review by rezk2ll — APPROVED** [review]
